File: src/cs/Samples/Samples/Samples/SolidEdge/Assembly/ReportInterference.cs
using ApiSamples.Samples.SolidEdge;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;

namespace ApiSamples.Samples.SolidEdge.Assembly
{
    /// <summary>
    /// Reports interference between all occurrences of the active assembly.
    /// </summary>
    class ReportInterference
    {
        static void RunSample(bool breakOnStart)
        {
            if (breakOnStart) System.Diagnostics.Debugger.Break();

            SolidEdgeFramework.Application application = null;
            SolidEdgeAssembly.AssemblyDocument assemblyDocument = null;
            SolidEdgeAssembly.Occurrences occurrences = null;
            SolidEdgeAssembly.Occurrence occurrence = null;
            SolidEdgeAssembly.InterferenceStatusConstants interferenceStatus;
            SolidEdgeConstants.InterferenceComparisonConstants compare = SolidEdgeConstants.InterferenceComparisonConstants.seInterferenceComparisonSet1vsAllOther;
            SolidEdgeConstants.InterferenceReportConstants reportType = SolidEdgeConstants.InterferenceReportConstants.seInterferenceReportPartNames;

            try
            {
                // Register with OLE to handle concurrency issues on the current thread.
                OleMessageFilter.Register();

                // Connect to or start Solid Edge.
                application = ApplicationHelper.Connect(true, true);

                // Get a reference to the active assembly document.
                assemblyDocument = application.TryActiveDocumentAs<SolidEdgeAssembly.AssemblyDocument>();

                if (assemblyDocument != null)
                {
                    // Get a reference to the Occurrences collection.
                    occurrences = assemblyDocument.Occurrences;

                    for (int i = 1; i <= occurrences.Count; i++)
                    {
                        // Get a reference to the occurrence.
                        occurrence = occurrences.Item(i);

                        Array set1 = Array.CreateInstance(occurrence.GetType(), 1);
                        object numInterferences = 0;
                        object retSet1 = Array.CreateInstance(typeof(SolidEdgeAssembly.Occurrence), 0);
                        object retSet2 = Array.CreateInstance(typeof(SolidEdgeAssembly.Occurrence), 0);
                        object confirmedInterference = null;
                        object interferenceOccurrence = null;

                        set1.SetValue(occurrence, 0);

                        // Check interference.
                        assemblyDocument.CheckInterference(
                            NumElementsSet1: set1.Length,
                            Set1: ref set1,
                            Status: out interferenceStatus,
                            ComparisonMethod: compare,
                            NumElementsSet2: 0,
                            Set2: Missing.Value,
                            AddInterferenceAsOccurrence: false,
                            ReportFilename: Missing.Value,
                            ReportType: reportType,
                            NumInterferences: out numInterferences,
                            InterferingPartsSet1: ref retSet1,
                            InterferingPartsOtherSet: ref retSet2,
                            ConfirmedInterference: ref confirmedInterference,
                            InterferenceOccurrence: out interferenceOccurrence,
                            IgnoreThreadInterferences: Missing.Value
                            );

                        // Process status.
                        switch (interferenceStatus)
                        {
                            case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusNoInterference:
                                break;
                            case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusConfirmedAndProbableInterference:
                            case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusConfirmedInterference:
                            case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusIncompleteAnalysis:
                            case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusProbableInterference:
                                if (retSet2 != null)
                                {
                                    for (int j = 0; j < (int)numInterferences; j++)
                                    {
                                        object obj1 = ((Array)retSet1).GetValue(j);
                                        object obj2 = ((Array)retSet2).GetValue(j);

                                        // Use ReflectionHelper class to get the object type.
                                        SolidEdgeFramework.ObjectType objectType1 = ReflectionHelper.GetObjectType(obj1);
                                        SolidEdgeFramework.ObjectType objectType2 = ReflectionHelper.GetObjectType(obj2);

                                        SolidEdgeFramework.Reference reference1 = null;
                                        SolidEdgeFramework.Reference reference2 = null;
                                        SolidEdgeAssembly.Occurrence occurrence1 = null;
                                        SolidEdgeAssembly.Occurrence occurrence2 = null;

                                        switch (objectType1)
                                        {
                                            case SolidEdgeFramework.ObjectType.igReference:
                                                reference1 = (SolidEdgeFramework.Reference)obj1;
                                                break;
                                            case SolidEdgeFramework.ObjectType.igPart:
                                            case SolidEdgeFramework.ObjectType.igOccurrence:
                                                occurrence1 = (SolidEdgeAssembly.Occurrence)obj1;
                                                break;
                                        }

                                        switch (objectType2)
                                        {
                                            case SolidEdgeFramework.ObjectType.igReference:
                                                reference2 = (SolidEdgeFramework.Reference)obj2;
                                                break;
                                            case SolidEdgeFramework.ObjectType.igPart:
                                            case SolidEdgeFramework.ObjectType.igOccurrence:
                                                occurrence2 = (SolidEdgeAssembly.Occurrence)obj2;
                                                break;
                                        }
                                    }
                                }
                                break;
                        }
                    }
                }
                else
                {
                    throw new System.Exception(Resources.NoActiveAssemblyDocument);
                }
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                OleMessageFilter.Unregister();
            }
        }
    }
}

File: src/vb/Samples/ApiSamples/Samples/Assembly/ReportInterference.vb
Imports SolidEdgeFramework.Extensions 'SolidEdge.Community.dll
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Reflection
Imports System.Text

Namespace ApiSamples.Assembly
	''' <summary>
	''' Reports interference between all occurrences of the active assembly.
	''' </summary>
	Friend Class ReportInterference
		Private Shared Sub RunSample(ByVal breakOnStart As Boolean)
			If breakOnStart Then
				System.Diagnostics.Debugger.Break()
			End If

			Dim application As SolidEdgeFramework.Application = Nothing
			Dim assemblyDocument As SolidEdgeAssembly.AssemblyDocument = Nothing
			Dim occurrences As SolidEdgeAssembly.Occurrences = Nothing
			Dim occurrence As SolidEdgeAssembly.Occurrence = Nothing
			Dim interferenceStatus As SolidEdgeAssembly.InterferenceStatusConstants = Nothing
			Dim compare As SolidEdgeConstants.InterferenceComparisonConstants = SolidEdgeConstants.InterferenceComparisonConstants.seInterferenceComparisonSet1vsAllOther
			Dim reportType As SolidEdgeConstants.InterferenceReportConstants = SolidEdgeConstants.InterferenceReportConstants.seInterferenceReportPartNames

			Try
				' Register with OLE to handle concurrency issues on the current thread.
				SolidEdgeCommunity.OleMessageFilter.Register()

				' Connect to or start Solid Edge.
				application = SolidEdgeCommunity.SolidEdgeInstall.Connect(True, True)

				' Get a reference to the active assembly document.
				assemblyDocument = application.GetActiveDocument(Of SolidEdgeAssembly.AssemblyDocument)(False)

				If assemblyDocument IsNot Nothing Then
					' Get a reference to the Occurrences collection.
					occurrences = assemblyDocument.Occurrences

					For i As Integer = 1 To occurrences.Count
						' Get a reference to the occurrence.
						occurrence = occurrences.Item(i)

						Dim set1 As Array = Array.CreateInstance(DirectCast(occurrence, Object).GetType(), 1)
						Dim numInterferences As Object = 0
						Dim retSet1 As Object = Array.CreateInstance(GetType(SolidEdgeAssembly.Occurrence), 0)
						Dim retSet2 As Object = Array.CreateInstance(GetType(SolidEdgeAssembly.Occurrence), 0)
						Dim confirmedInterference As Object = Nothing
						Dim interferenceOccurrence As Object = Nothing

						set1.SetValue(occurrence, 0)

						' Check interference.
						assemblyDocument.CheckInterference(NumElementsSet1:= set1.Length, Set1:= set1, Status:= interferenceStatus, ComparisonMethod:= compare, NumElementsSet2:= 0, Set2:= Missing.Value, AddInterferenceAsOccurrence:= False, ReportFilename:= Missing.Value, ReportType:= reportType, NumInterferences:= numInterferences, InterferingPartsSet1:= retSet1, InterferingPartsOtherSet:= retSet2, ConfirmedInterference:= confirmedInterference, InterferenceOccurrence:= interferenceOccurrence, IgnoreThreadInterferences:= Missing.Value)

						' Process status.
						Select Case interferenceStatus
							Case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusNoInterference
							Case SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusConfirmedAndProbableInterference, SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusConfirmedInterference, SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusIncompleteAnalysis, SolidEdgeAssembly.InterferenceStatusConstants.seInterferenceStatusProbableInterference
								If retSet2 IsNot Nothing Then
									For j As Integer = 0 To (DirectCast(numInterferences, Integer)) - 1
										Dim obj1 As Object = DirectCast(retSet1, Array).GetValue(j)
										Dim obj2 As Object = DirectCast(retSet2, Array).GetValue(j)

										' Use ReflectionHelper class to get the object type.
										Dim objectType1 As SolidEdgeFramework.ObjectType = ReflectionHelper.GetObjectType(obj1)
										Dim objectType2 As SolidEdgeFramework.ObjectType = ReflectionHelper.GetObjectType(obj2)

										Dim reference1 As SolidEdgeFramework.Reference = Nothing
										Dim reference2 As SolidEdgeFramework.Reference = Nothing
										Dim occurrence1 As SolidEdgeAssembly.Occurrence = Nothing
										Dim occurrence2 As SolidEdgeAssembly.Occurrence = Nothing

										Select Case objectType1
											Case SolidEdgeFramework.ObjectType.igReference
												reference1 = DirectCast(obj1, SolidEdgeFramework.Reference)
											Case SolidEdgeFramework.ObjectType.igPart, SolidEdgeFramework.ObjectType.igOccurrence
												occurrence1 = DirectCast(obj1, SolidEdgeAssembly.Occurrence)
										End Select

										Select Case objectType2
											Case SolidEdgeFramework.ObjectType.igReference
												reference2 = DirectCast(obj2, SolidEdgeFramework.Reference)
											Case SolidEdgeFramework.ObjectType.igPart, SolidEdgeFramework.ObjectType.igOccurrence
												occurrence2 = DirectCast(obj2, SolidEdgeAssembly.Occurrence)
										End Select
									Next j
								End If
						End Select
					Next i
				Else
					Throw New System.Exception(Resources.NoActiveAssemblyDocument)
				End If
			Catch ex As System.Exception
				Console.WriteLine(ex.Message)
			Finally
				SolidEdgeCommunity.OleMessageFilter.Unregister()
			End Try
		End Sub
	End Class
End Namespace
